ath79: add support for Ubiquiti NanoStation Loco M (XW)

rogerpueyo opened the pull request:
This commit adds support for the NanoStation Loco M2/M5 XW devices on the ath79 target (support was long ago available on ar71xx).

Specifications:

 - AR9342 SoC @535 MHz
 - 64 MB RAM
 - 8 MB SPI flash
 - 1x 10/100 Mbps Ethernet, 24 Vdc PoE-in
 - AR8032 switch
 - 2T2R 5 GHz radio, 22 dBm
 - 13 dBi built-in antenna
 - POWER/LAN green LEDs
 - 4x RSSI LEDs (red, orange, green, green)
 - UART (115200 8N1) on PCB

Flashing via TFTP:

 - Use a pointy tool (e.g., pen cap, paper clip) and keep the reset
   button on the device or on the PoE supply pressed
 - Power on the device via PoE (keep reset button pressed)
 - Keep pressing until LEDs flash alternatively LED1+LED3 =>
   LED2+LED4 => LED1+LED3, etc.
 - Release reset button
 - The device starts a TFTP server at 192.168.1.20
 - Set a static IP on the computer (e.g., 192.168.1.21/24)
 - Upload via tftp the factory image:
   $ tftp 192.168.1.20
   tftp> bin
   tftp> trace
   tftp> put openwrt-ath79-generic-xxxxx-ubnt_nanostation-loco-m-xw-squashfs-factory.bin

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

rogerpueyo committed:
ath79: add support for Ubiquiti NanoStation Loco M (XW)

This commit adds support for the NanoStation Loco M2/M5 XW devices
on the ath79 target (support was long ago available on ar71xx).

Specifications:

 - AR9342 SoC @535 MHz
 - 64 MB RAM
 - 8 MB SPI flash
 - 1x 10/100 Mbps Ethernet, 24 Vdc PoE-in
 - AR8032 switch
 - 2T2R 5 GHz radio, 22 dBm
 - 13 dBi built-in antenna
 - POWER/LAN green LEDs
 - 4x RSSI LEDs (red, orange, green, green)
 - UART (115200 8N1) on PCB

Flashing via TFTP:

 - Use a pointy tool (e.g., pen cap, paper clip) and keep the reset
   button on the device or on the PoE supply pressed
 - Power on the device via PoE (keep reset button pressed)
 - Keep pressing until LEDs flash alternatively LED1+LED3 =>
   LED2+LED4 => LED1+LED3, etc.
 - Release reset button
 - The device starts a TFTP server at 192.168.1.20
 - Set a static IP on the computer (e.g., 192.168.1.21/24)
 - Upload via tftp the factory image:
   $ tftp 192.168.1.20
   tftp> bin
   tftp> trace
   tftp> put openwrt-ath79-generic-xxxxx-ubnt_nanostation-loco-m-xw-squashfs-factory.bin

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

adrianschmutzler commented:
I'm somewhat unhappy with the name "ubnt_nanostation-loco-m-xw".
I admit that this is the proper solution.
However, it's long and not really nice. But I'm also not sure about using ubnt,loco-m-xw, as it's technically not the correct name.
How do the others feel about that?

rogerpueyo commented:
I've also backported it to openwrt-19.07 at https://github.com/rogerpueyo/openwrt/tree/ath79-ubnt-nanostation-m-loco-xw-19.07.

adrianschmutzler commented:
@rogerpueyo Typically device support is not backported.
Since 19.07 contains a working ar71xx image for the loco-m-xw, changes for backporting ath79 are even lower. So, I would not invest time in maintaining that backport, unless you want to use it yourself.

rogerpueyo commented:
> @rogerpueyo Typically device support is not backported.
> Since 19.07 contains a working ar71xx image for the loco-m-xw, changes for backporting ath79 are even lower. So, I would not invest time in maintaining that backport, unless you want to use it yourself.

Cool, no worries :)

ynezz commented:
> I'm somewhat unhappy with the name "ubnt_nanostation-loco-m-xw".
> I admit that this is the proper solution.

UBNT should make product names shorter :) Putting jokes aside, just use what vendor uses to lower the end user confusion.

adrianschmutzler commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -108,6 +108,14 @@ define Device/ubnt_rocket-m
+define Device/ubnt_nanostation-loco-m-xw
+  $(Device/ubnt-xw)
+  DEVICE_MODEL := Nanostation Loco M
+  DEVICE_VARIANT := XW
```
Since ynezz just pulled my PR #2425 , you can remove this line as it will be inherited from ubnt-xw.

rogerpueyo replied:
Line "DEVICE_VARIANT := XW" removed.

rogerpueyo commented:
> UBNT should make product names shorter :) Putting jokes aside, just use what vendor uses to lower the end user confusion.

Indeed, they use different names on the devices' datasheet :man_facepalming:, but "Nanostation Loco M{2, 5}" seems to be the most consistent one, plus the internal XW version.

Fixed requested changes and rebased master :+1:

rogerpueyo commented:
Rebased.

rogerpueyo commented:
Rebased master.

psyborg55 commented:
> I'm somewhat unhappy with the name "ubnt_nanostation-loco-m-xw".
> I admit that this is the proper solution.
> However, it's long and not really nice. But I'm also not sure about using ubnt,loco-m-xw, as it's technically not the correct name.
> How do the others feel about that?

how about "ubnt_ns-loco-m-xw" ?

rogerpueyo commented:
@psyborg55 I see your point about the naming; the ar71xx target used all sort of abbreviations like "nano", "lbe", "rs"... However, the ath79 one mostly employs the whole name ("nanostation", "nanobeam", "routerstation", etc.). It may be long and nasty, but "nanostation-m-loco-xw" leaves no room for confusion, while "ns" could be misleading and eventually match a future product's acronym.

adrianschmutzler commented:
I also have made peace with nanostation-loco-m-xw now ...

CodeFetch commented:
It's a bit OT, but I'd like to make you aware of a problem I see with Ubiquiti AirMax XM/XW device support in OpenWrt.

The Wiki says:
```
The bullet-m image is the base of the airmax series.
The rocket-m image is the bullet-m with an additionnal usb driver.
The nano-m image is the bullet-m with an additionnal ethernet.
The NanoBridge-m works fine with bullet-m r30919
```

This is how it was handled with ar71xx.

I've tried to find out the PA and antenna gains of AirMax devices so that we could introduce this in DTS. But this would mean that we need to introduce a lot more different devices.

What do you think is the correct way to do this? Either we could create one DTS for every AirMax M device (see list below) or we could follow the tradition of ar71xx and introduce a script which sets the proper values using iw depending on the model.

Here is an overview of my findings so far:
```
		'e112' = { -- XM
			pa_gain = 10
		},
		'e115' = { -- XM
			pa_gain = 3
		},
		'e2c4' = { -- XM
			pa_gain = 6
		}
		'e2d4' = { -- XM
			pa_gain = 6
		}
		'e2e5' = { -- XM
			pa_gain = 4
		}
		'e305' = {
			pa_gain = 6
		},
		'e3a5' = { -- XM
			pa_gain = 5
		}
		'e3b5' = { -- XM / XW
			pa_gain = 6
		}
		'e402' = { -- XM
			pa_gain = 10
		}
		'e405' = { -- XM
			pa_gain = 1
		}
		'e4a5' = { -- AirRouter? XM
			pa_gain = 1
		},
		'e5e5' = { -- Powerbeam? XW
			pa_gain = 5
		},
		'e6a2' = { -- XW
			pa_gain = 1
		},
		'e7f8' = { -- XW
			pa_gain = 2
		},
		'e875' = { -- XW
			pa_gain = 4
		},
		'e879' = { -- XW
			pa_gain = 2
		},
		'e895' = { -- Powerbeam M5? XW
			pa_gain = 4
		},
		'e8b5' = {
			pa_gain = 5
		},

		-- NanoStation M2
		-- ERP 28 dBm according to datasheet
		'e002' = { -- ??? (EU?) XM
			pa_gain = 6
			ant_gain = 11
		},
		'e012' = { -- ??? (US?) XM
			pa_gain = 10
			ant_gain = 11
		},
		'e866' = { -- XW
			pa_gain = 6
			ant_gain = 11
		},

		-- NanoStation M3? Loco?
		-- ERP 25 dBm according to datasheet
		'e003' = { -- XM
			pa_gain = 3
			ant_gain = 13
		}
		-- NanoStation M3
		-- ERP 25 dBm according to datasheet
		'e035' = {
			pa_gain = 3
			ant_gain = 13
		}

		-- NanoStation M5
		-- ERP 27 dBm according to datasheet
		'e005' = { -- ??? (EU?) XM
			pa_gain = 5
			ant_gain = 16
		},
		'e805' = { -- ??? (US?) XM
			pa_gain = 5
			ant_gain = 16
		},
		'e855' = { -- XW
			pa_gain = 5
			ant_gain = 16
		},

		-- NanoStation M365?
		-- ERP 25 dBm according to datasheet
		'e006' = { -- XM
			pa_gain = 5
			ant_gain = 13
		}

		-- NanoStation Loco M9 XM
		-- ERP 28 dBm according to datasheet
		'e009' = {
			pa_gain = 6
			ant_gain = 8
		}

		-- NanoStation Loco M2
		-- ERP 23 dBm according to datasheet
		'e0a2' = { -- XM
			pa_gain = 2
			ant_gain = 9
		},
		'e867' = { -- XW
			pa_gain = 2
			ant_gain = 9
		},

		-- NanoStation Loco M5
		-- ERP 23 dBM according to datasheet
		'e0a5' = { -- XM ??? (EU?)
			pa_gain = 1
			ant_gain = 13
		},
		'e8a5' = { -- XM ??? (US?)
			pa_gain = 1
			ant_gain = 13
		},
		'e845' = { -- XW
			pa_gain = 1
			ant_gain = 13
		},

		-- Rocket M2
		-- ERP 28 dBm according to datasheet
		'e102' = { -- ??? (EU?)
			pa_gain = 6
			ant_detach = true
		},
		'e1b2' = { -- HP (US?)
			pa_gain = 10
			ant_detach = true
		},
		'e1c2' = { -- GPS
			pa_gain = 10
			ant_detach = true
		},
		'e1d2' = { -- Titanium
			pa_gain = 10
			ant_detach = true
		},
		'e868' = { -- XW
			pa_gain = 7
			ant_detach = true
		},

		-- Rocket M5
		-- ERP 27 dBm according to datasheet
		'e105' = { -- ??? (EU?)
			pa_gain = 5
			ant_detach = true
		},
		'e1b5' = { -- ??? (US?)
			pa_gain = 5
			ant_detach = true
		},
		'e1c5' = { -- GPS
			pa_gain = 5
			ant_detach = true
		},
		'e1d5' = { -- Titanium
			pa_gain = 5
			ant_detach = true
		},
		'e1e5' = { -- ???
			pa_gain = 5
			ant_detach = true
		},
		'e6b5' = { -- XW
			pa_gain = 5
			ant_detach = true
		},

		-- PowerBridge M3
		-- ERP 25 dBm according to datasheet
		'e1a3' = {
			pa_gain = 3
			ant_gain = 20
		},

		-- PowerBridge M5
		-- ERP 27 dBm according to datasheet
		'e1a5' = {
			pa_gain = 5
			ant_gain = 25
		},

		-- Rocket M3
		-- ERP 25 dBm according to datasheet
		'e1b3' = { -- XM
			pa_gain = 3
			ant_detach = true
		},
		'e1c3' = { -- GPS?
			pa_gain = 3
			ant_detach = true
		},
		'e1d3' = { -- Titanium?
			pa_gain = 3
			ant_detach = true
		},
		'e1e3' = { -- ???
			pa_gain = 3
			ant_detach = true
		},

		-- Rocket M365
		-- ERP 25 dBm according to datasheet
		'e1b6' = { -- XM
			pa_gain = 5
			ant_detach = true
		},

		-- Rocket M900
		-- ERP 28 dBm according to datasheet
		'e1b9' = { -- XM
			pa_gain = 6
			ant_detach = true
		},
		'e1d9' = { -- Titanium?
			pa_gain = 6
			ant_detach = true
		},

		-- Bullet M2
		-- ERP 28 dBm according to datasheet
		'e202' = { -- BM2HP
			pa_gain = 12
			ant_detach = true
		},
		'e2d2' = { -- BM2-Ti (Titanium)
			pa_gain = 12
			ant_detach = true
		},
		'e2c3' = { -- XW
			pa_gain = 6
			ant_detach = true
		},

		-- Bullet M5 XM
		-- ERP 25 dBm according to datasheet
		'e205' = { -- BM5HP
			pa_gain = 6
			ant_detach = true
		},
		'e2d5' = { -- BM5-Ti (Titanium)
			pa_gain = 6
			ant_detach = true
		},

		-- AirGrid M2 XM
		-- FIXME: There must be four versions
		--    (AG-HP-2G16, AG-HP-2G20, AG2G16, AG2G20)
		-- ERP 20 dBm according to datasheet
		-- HP: ERP 28 dBm according to datasheet
		'e212' = {
			pa_gain = 1
			ant_gain = 20
		},

		-- AirGrid M5
		-- FIXME: There must be four versions
		--    (AG-HP-5G23, AG-HP-5G27, AG5G23, AG5G27)
		-- ERP 20 dBm according to datasheet
		-- HP: ERP 25 dBm according to datasheet
		'e215' = { -- XM
			pa_gain = 1
			ant_gain = 27
		},
		'e835' = { -- XW HP?
			pa_gain = 6
			ant_gain = 27
		},

		-- NanoBridge M2 XM 
		-- ERP 23 dBm according to datasheet
		'e232' = { -- NB-2G18?
			pa_gain = 2
			ant_gain = 18
		},
		'e2b2' = { -- ??? (US?)
			pa_gain = 10
			ant_gain = 18
		},

		-- NanoBridge M3
		-- FIXME: There might be a NBM365
		-- ERP 25 dBm according to datasheet
		'e233' = {
			pa_gain = 3
			ant_gain = 22
		},
		'e2a3' = { -- TODO: Check
			pa_gain = 3
			ant_gain = 22
		},

		-- NanoBridge M5 XM
		-- FIXME: One must be NBM5-22 and the other NBM5-25
		-- ERP 23 dBm according to datasheet
		'e235' = {
			pa_gain = 1
			ant_gain = 25
		},
		'e2b5' = {
			pa_gain = 1
			ant_gain = 25
		},

		-- NanoBridge M9?
		-- ERP 28 dBm according to datasheet
		'e2b9' = {
			pa_gain = 6
			ant_gain = 11
		},

		-- PicoStation M2
		-- ERP 28 dBm according to datasheet
		'e302' = { -- ??? (US?)
			pa_gain = 12
			ant_gain = 2 -- FIXME: US has 5 dBi antenna
			ant_detach = true
		},

		-- AirRouter (EU) (XM)
		-- ERP 19 dBm according to datasheet
		'e4a2' = {
			-- This is an important hint as WLAN router
			-- manufacturers often use 1 dBm as safeguard
			-- The WLAN chip's ERP is 20 dBm. 20 - 1 = 19
			-- FIXME: antenna gain unknown
			pa_gain = 1
		},

		-- AirRouter HP (US) (XM)
		-- ERP 28 dBm according to datasheet
		'e4b2' = {
			-- FIXME: antenna gain unknown 
			pa_gain = 9
		},

		-- NanoBeam M2
		-- ERP 28 dBm according to datasheet
		'e2c2' = { -- ???
			-- FIXME: antenna gain unknown
			pa_gain = 10
			ant_gain = 13
		},
		'e812' = { -- NBE-M2-13
			pa_gain = 6
			ant_gain = 13
		},

		-- WispStation M5
		-- ERP 23 dBm according to datasheet
		'e345' = {
			pa_gain = 6
		}

		-- PowerBeam M5
		-- ERP 26 dBm according to datasheet
		'e3e5' = { -- PBE-M5-300
			pa_gain = 4
			ant_gain = 22
		},
		'e4e5' = { -- PBE-M5-400
			pa_gain = 4
			ant_gain = 25
		},
		'e6e5' = { -- PBE-M5-400 ISO!!
			pa_gain = 4
			ant_gain = 25
		},
		'e885' = { -- PBE-M5-620
			pa_gain = 4
			ant_gain = 29
		},

		-- Powerbeam M2 400?
		-- ERP 28 dBm according to datasheet
		'e6c2' = { -- PBE-M2-400
			pa_gain = 6
			ant_gain = 18 dBi
		},

		-- NanoBeam M5
		-- ERP 26 dBm according to datasheet
		'e815' = { -- NBE-M5-16
			pa_gain = 4
			ant_gain = 16
		},
		'e825' = { -- NBE-M5-19
			pa_gain = 4
			ant_gain = 19
		},

		-- LiteBeam M5 (XW)
		-- ERP 25 dBm according to datasheet
		'e865' = { -- LBE-M5-23
			pa_gain = 6
			ant_gain = 23
		},
```

psyborg55 commented:
since the whole target is about DTS i think it would be preferred way to do it, rather than further complicate things with userspace scripts

CodeFetch commented:
@psyborg55 Okay, but that would mean that one needs a separate DTS for NanoStation Loco M2 XW and M5 XW here:

NanoStation Loco M2:
pa_gain = 2
ant_gain = 9

and NanoStation Loco M5
pa_gain = 1
ant_gain = 13

That's why I'm asking. I've thought DTS would be the preferred way, too, but it would introduce a lot more models/DTS files and consumes more disk space for generated images. And maybe some people are unsure which device they have. E.g. there are two versions of the NanoBridge M2 XM which can't be differentiated easily.
Okay, I think I've answered my question myself...

ynezz commented:
> I've tried to find out the PA and antenna gains of AirMax devices so that we could introduce this in DTS.

How is vendor handling it? Hardcoding those values into the DTS doesnt make sense, driver in the DTS should be pointed to the location of this information in the flash/eeprom and use it. If unsure, you can always ask upstream on linux-wireless mailing list.

psyborg55 commented:
> Okay, but that would mean that one needs a separate DTS for NanoStation Loco M2 XW and M5
> XW here

yes, separate DTS for each device. i don't see a problem with it. one of the reason DTS target was created was to reduce final image size, which happens by including only device's specific DTS/DTB in its image.

CodeFetch commented:
> How is vendor handling it? Hardcoding those values into the DTS doesnt make sense, driver in the DTS should be pointed to the location of this information in the flash/eeprom and use it. If unsure, you can always ask upstream on linux-wireless mailing list.

Ubiquiti has a kernel module (HAL) which reads the subvendor device ID of the WLAN module ([found in flash](https://github.com/openwrt/openwrt/blob/845b2a1cfe307394f3f2144cce085bbb5e171ecc/target/linux/ar71xx/base-files/lib/ar71xx.sh#L71)). The kernel module contains the PA gains for all XM/XW devices respectively. Frankly I haven't found out where Ubiquiti saves the antenna gain (possibly in ath9k's ART).

adrianschmutzler commented:
@ynezz Note that this is actually an important topic, because if nothing changed recently, the antenna_gain for Ubiquiti devices is just not set by OpenWrt, so all devices use illegally high TX power by default (and continue to do so if not adjusted manually).

@CodeFetch I don't think it's reasonable to create a DTS for every device in that list, even if we merge the ones with same values. Even if we do not care for additional build time, it will make the target pretty unmaintainable. ynezz' suggestion sounds interesting if we can manage to get it working. Otherwise, I'd go for a userspace script.

psyborg55 commented:
> The kernel module contains the PA gains for all XM/XW devices respectively. Frankly I haven't
> found out where Ubiquiti saves the antenna gain (possibly in ath9k's ART)

unless they're both stored in ART that suggestion doesn't make much sense.

>  it will make the target pretty unmaintainable

not really valid argument. you set these values once and don't touch them anymore. if some of the devices need different values for other registers you can also simply reuse specific DTS file

adrianschmutzler commented:
> > it will make the target pretty unmaintainable
> 
> not really valid argument. you set these values once and don't touch them anymore. if some of the devices need different values for other registers you can also simply reuse specific DTS file

You have to keep in mind that you still will need to duplicate the entries in other file, too: 01_leds, 02_network, ath10k, generic-ubnt.mk, ...

I don't say that this is a bad idea in general, but if you really want to account for all antenna_gain differences, it will become a lot of devices. I don't say that this is impossible, but I do say that this a severe drawback.

psyborg55 commented:
well, maybe, unless you can come up with some DTS include solution. but having few dozen lines more of code in sources should not pose a restriction for proper device support (especially not if target does what was promised and doesn't increase final image size due to such changes).
looking at other devices in these sources they all have device-unique entries (those with identical settings grouped together), only ubnt devices seem to need conversion to proper DTS format

adrianschmutzler commented:
@rogerpueyo Please rebase onto base-files split.
I think the Ubiquiti DTS discussion is general, and despite the rebase this should be reviewed in its current state.

rogerpueyo commented:
> @rogerpueyo Please rebase onto base-files split.
> I think the Ubiquiti DTS discussion is general, and despite the rebase this should be reviewed in its current state.

Rebased and pushed, thanks @adrianschmutzler.

adrianschmutzler commented:
@ynezz @981213 @blocktrron 
Can one of you have a look at the ethernet setup here? Everything else should be fine.

References:
nano-m-xw @ ar71xx
https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-ubnt-xm.c#L472

loco-m-xw @ ar71xx
https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-ubnt-xm.c#L497

adrianschmutzler commented:
Despite, I have a stupid question:

In ar934x.dtsi, spi is set up as follows:
```
spi: spi@1f000000 {
   ...
   reg = <0x1f000000 0x1c>;
   ...
}
```

In ar9342_ubnt_xw.dtsi, we have
```
			eeprom: partition@ff0000 { # This seems to be a copy-paste error from ubnt-wa
				label = "EEPROM";
				reg = <0x7f0000 0x010000>;
				read-only;
			};
```

and

```
&wmac {
	status = "okay";

	mtd-cal-data = <&eeprom 0x1000>;
};
```

But in ar71xx mach-files, there is:
```
u8 *eeprom = (u8 *) KSEG1ADDR(0x1fff0000);
...
#define UAP_PRO_WMAC_CALDATA_OFFSET	0x1000
...
ath79_register_wmac(eeprom + UAP_PRO_WMAC_CALDATA_OFFSET, NULL);
```

This would be 0x1f7f1000 on ath79 and 0x1fff1000 on ar71xx. However, it seems like both device families receive the correct caldata. So, where is my error?

981213 commented:
> This would be 0x1f7f1000 on ath79 and 0x1fff1000 on ar71xx. However, it seems like both device families receive the correct caldata. So, where is my error?

Ah... I‘m not sure how to explain this behavior in English and here's my attempt :P

When receiving a reading request with addresses beyond its capacity, SPI NOR will ignore the exceeded higher bits and return corresponding data.
e.g. If we have an 8M flash, the max valid address is 0x7fffff and 23rd and higher bits are ignored.
In these QCA chips, bootloader will configure SPI controller to map 16MB of SPI NOR starting at 0x1f000000. Reading KSEG1ADDR(0x1fff1000) is telling NOR flash to return data after 0xff0000 and if we are on an 8M flash it will interpret this as 0xff1000 & 0x7fffff = 0x7f1000

adrianschmutzler commented:
@981213 Thanks for enlightening me. So, that means using KSEG1ADDR(0x1f7f1000) would just give the same result? (Don't worry, I won't attempt to change this for ar71xx ...).

(What should be change are the labels in ath79 xw: partition@ff0000 -> partition@7f0000)

dangowrt commented:
Any news on this? I'm around a bunch of nanostation-loco-m-xw and could do testing, merge if ready, backport to 19.07, ...
@rogerpueyo @ynezz

adrianschmutzler commented on target/linux/ath79/dts/ar9342_ubnt_nanostation-loco-m-xw.dts:
```diff
@@ -0,0 +1,30 @@
+
+	phy-mask = <0x1>;
+
+	phy4: ethernet-phy@4 {
```
After what I've learned with the mikrotik PR, this should be
`phy1: ethernet-phy@1 {` ?

rogerpueyo replied:
It should. I haven't tested it on the real hardware but it's just a matter of naming, isn't it?

adrianschmutzler replied:
I understand it the same way, it should be purely cosmetical.

adrianschmutzler commented on target/linux/ath79/dts/ar9342_ubnt_nanostation-loco-m-xw.dts:
```diff
@@ -0,0 +1,30 @@
+&eth0 {
+	status = "okay";
+
+	phy-mode = "mii";
```
That's already set in ath79.dtsi, so it can be removed here.

rogerpueyo replied:
:+1:

adrianschmutzler commented on target/linux/ath79/dts/ar9342_ubnt_nanostation-loco-m-xw.dts:
```diff
@@ -0,0 +1,30 @@
+	status = "okay";
+
+	phy-mode = "mii";
+	phy-handle = <&phy4>;
```
Update to phy1?

adrianschmutzler commented on target/linux/ath79/dts/ar9342_ubnt_nanostation-loco-m-xw.dts:
```diff
@@ -0,0 +1,30 @@
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0x1>;
```
I have no idea about this one.

adrianschmutzler replied:
Ah, this one is correct:
`ath79_eth0_data.phy_mask = BIT(1);`

adrianschmutzler commented:
Okay, I checked the ar71xx mach files again and it seems that this PR should be fine with the requested (cosmetic) changes above.

There is still something I wonder about, though:
```
static struct mdio_board_info ubnt_loco_m_xw_mdio_info[] = {
	{
		[...]
		.platform_data = &ubnt_loco_m_xw_at803x_data,
	},
};
static struct at803x_platform_data ubnt_loco_m_xw_at803x_data = {
	.has_reset_gpio = 1,
	.reset_gpio = 0,
};
```
Similar settings as in the second struct are only present in ar71xx for devices not moved to ath79 yet (including the Loco M XW, but none of the other devices in mach-ubnt-xm.c). I do not know whether this should be accounted for anywhere.

rogerpueyo commented:
> Any news on this? I'm around a bunch of nanostation-loco-m-xw and could do testing, merge if ready, backport to 19.07, ...
> @rogerpueyo @ynezz

Sorry @dangowrt, @adrianschmutzler, but I don't have any Loco M XW around anymore. :(((

I have asked around for one, but it might take some weeks until I can have my hands on it. If you can make any progress in the meantime, please, proceed. I've fixed Adrian's suggestions for cosmetic changes.

adrianschmutzler commented:
@981213 Can you comment on my question two posts above (at803x_platform_data)?

adrianschmutzler commented:
@rogerpueyo I assume you tested in the first place? Since the changes requested by me are cosmetical, if it worked then it should still work. So I wouldn't consider that a blocker for merging.

rogerpueyo commented:
@adrianschmutzler Yes, I tested it on the actual device the last change until today's and, as far as I can remember, everything was working. But if you want to wait for somebody else's tests, I'd understand it. @dangowrt said he had a few devices where he could do some testing.

rogerpueyo commented:
@adrianschmutzler I got one of these devices, quicker than expected. I rebased master and tested the image. It works as expected so, if there are no other issues, I guess it's ready to merge.

adrianschmutzler commented:
Thanks, I've merged it. I think I will ask for the mdio-reset topic separately on the list at some point.

I would also backport this to 19.07, as backporting is quite simple and it will complete the Ubiquiti family there. Can you please run-test with my staging-19.07 tree:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging-19.07

adrianschmutzler commented:
http://lists.infradead.org/pipermail/openwrt-devel/2020-January/021530.html

adrianschmutzler commented:
Ping: Please test my staging tree.

dangowrt commented:
I'm afraid having merged this without the MDIO PHY reset GPIO will make people climb on poles (once the initially think everything works fine and then after a couple of days end up with RX-deaf Ethernet PHY...)

adrianschmutzler commented:
Thanks for reminding me, I've totally forgotten about this. I have reverted the commit in 19.07.

rogerpueyo commented:
So, is this what was missing? https://github.com/rogerpueyo/openwrt/commit/65d7d0950f1d97e1b6fa9a817c7d5028a296edef#diff-f52ac4d68c55bd745059711ab889f258R19
How can we test that the reset-gpios thing is actually working?
